src/associations/has-many.js

```diff
@@ -346,16 +346,25 @@ class HasMany extends Association {
       update[this.foreignKey] = null;
 
       updateWhere = {
-        [this.target.primaryKeyAttribute]: obsoleteAssociations.map(associatedObject => associatedObject[this.target.primaryKeyAttribute]),
+        [this.target.primaryKeyAttribute]: obsoleteAssociations.map(
+          associatedObject => associatedObject[this.target.primaryKeyAttribute],
+        ),
       };
 
-      promises.push(this.target.unscoped().update(
-        update,
-        {
-          ...options,
-          where: updateWhere,
-        },
-      ));
+      // update or delete obsolete associations depending on foreign key
+      let udMethod;
+      const udArgs = [{ ...options, where: updateWhere }];
+      const foreignKeyIsNullable
+        = this.target.rawAttributes[this.identifierField].allowNull;
```

Comment:
`identifierField` is a column name, `rawAttributes` is indexed by attribute name. I think this should be

```suggestion
        = this.target.rawAttributes[this.foreignKey].allowNull;
```

I don't think `allowNull` is normalized, so it could be undefined too (which defaults to true). The solution should be:

```suggestion
        = this.target.rawAttributes[this.foreignKey].allowNull !== false;
```